### PR TITLE
Experiment ac rust/v11

### PR DESCRIPTION
**Pull request by victorjulien**
replace #9904, includes #9908

Implement "ac-rs" using the aho-corasick Rust crate. Sadly, it doesn't give the performance I hoped for. Don't see more ways to optimize it.

### Intel Xeon E5-2680v4
```
profile-mpm-sandnet.pcap-open.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142              942        4209906         25619         10.3b  37.61 
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141              945       20108490         53608         21.5b  45.88 
profile-mpm-sandnet.pcap-open.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142              882         832744         23083          9.3b  36.24 
profile-mpm-sandnet.pcap-open.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143              942        3383613         12789          5.1b  28.60 
profile-mpm-sandnet.pcap-pro.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140              902        1086180         34368         13.8b  38.75 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140              951       19817724         57077         22.9b  43.69 
profile-mpm-sandnet.pcap-pro.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141              900        1482510         30627         12.3b  37.36 
profile-mpm-sandnet.pcap-pro.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142              960         888276         20900          8.4b  33.08
```
HS almost 3x AC-RS, AC almost 2x AC-RS.

### AMD Threadripper 2990wx
```
profile-mpm-sandnet.pcap-open.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1320        2114310         24750          9.9b  33.05 
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140             1350        2273430         56612         22.7b  42.61 
profile-mpm-sandnet.pcap-open.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1350        2251170         24463          9.8b  32.14 
profile-mpm-sandnet.pcap-pro.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401139             1350       12749880         40299         16.2b  34.29 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1350        4107720         68919         27.6b  40.23 
profile-mpm-sandnet.pcap-pro.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401144             1320        2062860         38447         15.4b  32.51
```
AC almost 2x

### Intel Xeon W-2245
```
profile-mpm-sandnet.pcap-open.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1078         725199         29635         11.9b  34.61 
profile-mpm-sandnet.pcap-open.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1088        2425744         28189         11.3b  33.55 
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401146             1111        2915524         70305         28.2b  44.42 
profile-mpm-sandnet.pcap-open.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1252         504847         15159          6.1b  26.25 
profile-mpm-sandnet.pcap-pro.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1083        2541907         42546         17.1b  35.06 
profile-mpm-sandnet.pcap-pro.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1100        2530907         38909         15.6b  33.28 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140             1083        2829304         76386         30.6b  41.15 
profile-mpm-sandnet.pcap-pro.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401144             1228        1248314         24381          9.8b  29.02
```
HS is 3x the perf, AC 2x.

### CODE REVIEW

File: rust/src/mpm_ac_rs.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+/// Search for the patterns. Returns number of matches.
+/// Per pattern found sids are only appended once.
+#[no_mangle]
+pub unsafe extern "C" fn rs_mpm_acrs_search(state: &AhoCorasickState, data: *const u8, data_len: u32,
+    cb: unsafe extern "C" fn(*mut std::os::raw::c_void, *const u32, u32),
+    cbdata: *mut std::os::raw::c_void) -> u32
+{
+    let haystack = build_slice!(data, data_len as usize);
+    let mut match_cnt : u32 = 0;
+    // array of bools for patterns we found
```
Comment (victorjulien) [inline rust/src/mpm_ac_rs.rs:123]: comment is wrong, it's used as a bitarray now

**Comment by victorjulien** [comment]
Updated numbers after setting "kind" to DFA based on https://github.com/BurntSushi/aho-corasick/discussions/136#discussioncomment-7695202

### Intel Xeon W-2245
```
profile-mpm-sandnet.pcap-open.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1078         725199         29635         11.9b  34.61 
profile-mpm-sandnet.pcap-open.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1088        2425744         28189         11.3b  33.55 
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140             1083        2492467         39820         16.0b  39.83 
profile-mpm-sandnet.pcap-open.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1252         504847         15159          6.1b  26.25 
profile-mpm-sandnet.pcap-pro.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1083        2541907         42546         17.1b  35.06 
profile-mpm-sandnet.pcap-pro.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1100        2530907         38909         15.6b  33.28 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1073        1325263         57487         23.1b  40.56 
profile-mpm-sandnet.pcap-pro.rules-hs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401144             1228        1248314         24381          9.8b  29.02 
```
So in open, this improves things from 28.2b to 16b ticks.
In pro from 30.6b ticks to 23.1b. However still considerably slower than the others.

### AMD Threadripper 2990wx
```
profile-mpm-sandnet.pcap-open.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1320        2114310         24750          9.9b  33.05 
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1350       10578090         35419         14.2b  38.42 
profile-mpm-sandnet.pcap-open.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1350        2251170         24463          9.8b  32.14 
profile-mpm-sandnet.pcap-pro.rules-ac-ks/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401139             1350       12749880         40299         16.2b  34.29 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1350       11825370         57434         23.0b  39.84 
profile-mpm-sandnet.pcap-pro.rules-ac/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401144             1320        2062860         38447         15.4b  32.51 
```
Open: 22.7b -> 14.2b 
Pro: 27.6b -> 23b
So again, a good improvement but still slower than what we already have.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 181 | 191 | 105.52% | 

Pipeline 16764

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 16765

**Comment by victorjulien** [comment]
Direct DFA calls attempt has the following results
### Intel Xeon W-2245
```
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401143             1135        2724221         39536         15.9b  39.57 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401140             1074        2606235         56937         22.8b  40.66 
```
So a small improvement over the last attempt.

### AMD Threadripper 2990wx
```
profile-mpm-sandnet.pcap-open.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401141             1350       10436310         34459         13.8b  38.41 
profile-mpm-sandnet.pcap-pro.rules-ac-rs/packet_stats.log:PROF_DETECT_PF_PAYLOAD      IPv4       6        401142             1350       10303920         56272         22.6b  39.30 
```
Also a small improvement, but still not where we'd hope things to land.

**Comment by victorjulien** [comment]
Explicitly disabling the prefilter feature seems to have a minimal effect.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 181 | 192 | 106.08% | 

Pipeline 16777

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 16779

**Comment by victorjulien** [comment]
Performance wasn't there to justify doing more work on this.

**Pull request closed without merging**